### PR TITLE
Bump opencv/4.8.1

### PR DESCRIPTION
**Pull request by SpaceIm**
There were several modifications in OpenCV build system since 4.5.5.
The most noticeable ones which can be seen in this PR:

- `dnn` module:
  - optionally depends on flatbuffers since 4.8.0 (vendored header-only flavor), so `with_flatbuffers` option has been added, enabled by default.
  - protobuf is not mandatory anymore since 4.8.0, so `with_protobuf` option has been added, enabled by default
- `highgui` module:
  - optional wayland backend since 4.7.0, and therefore `with_wayland` option has been added (disabled by default)
- `imgcodecs` module:
  - add `with_avif` option (disabled by default) since 4.8.0 to enable AVIF support through libavif.
  - Since 4.7.0, `libspng` can be used as a replacement of `libpng` for PNG backend, it's left as an exercise for another PR since it would lead to a breaking change of `with_png` option (from `[True, False]` to `[False, "libpng", "libspng"]`).
- `aruco` extra module: depends on `objdetect` since 4.7.0
- `barcode` extra module : removed in 4.8.0, this code lives into `objdetect` module now.
- `cudacodec` extra module : depends on `cudaarithm` & `cudawarping` instead of `cudev` since 4.7.0 (I don't know if it's a build fix or a consequence of a code modification).
- `wechat_qrcode` extra module: depends on `objdetect` since 4.7.0

2 patches of 4.5.5 are not required since 4.6.0 (4.5.5-0002-objdetect-without-dnn.patch and 4.5.1-0002-alphamat-header.patch), but others are still needed in order to ensure robust build.

closes https://github.com/conan-io/conan-center-index/issues/21002

---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by SpaceIm** [comment]
Also I believe that ffmpeg >= 5 is still not supported by OpenCV 4.8 (but I've not tested). The latest update is https://github.com/opencv/opencv/pull/23846 and it's still pointing to ffmpeg 4.x

**Comment by SpaceIm** [comment]
And I forgot to say that OpenVINO is supported since 4.6.0 in `dnn` & `gapi` modules, but not handled by this PR (it's forced disabled). Once https://github.com/conan-io/conan-center-index/pull/19164 merged, an option could be added in another PR.

**Comment by SpaceIm** [comment]
```
/home/conan/w/prod-v1/BuildSingleReference/.conan/data/freetype/2.13.0/_/_/package/bf05ac88aac1986abd35f257eeb0064299824400/lib/libfreetype.a(sfnt.c.o): In function `sfnt_init_face':
sfnt.c:(.text+0xc068): undefined reference to `BrotliDecoderDecompress'
/home/conan/w/prod-v1/BuildSingleReference/.conan/data/freetype/2.13.0/_/_/package/bf05ac88aac1986abd35f257eeb0064299824400/lib/libfreetype.a(ftbzip2.c.o): In function `ft_bzip2_stream_close':
ftbzip2.c:(.text+0x79): undefined reference to `BZ2_bzDecompressEnd'
/home/conan/w/prod-v1/BuildSingleReference/.conan/data/freetype/2.13.0/_/_/package/bf05ac88aac1986abd35f257eeb0064299824400/lib/libfreetype.a(ftbzip2.c.o): In function `ft_bzip2_file_fill_output':
ftbzip2.c:(.text+0x183): undefined reference to `BZ2_bzDecompress'
/home/conan/w/prod-v1/BuildSingleReference/.conan/data/freetype/2.13.0/_/_/package/bf05ac88aac1986abd35f257eeb0064299824400/lib/libfreetype.a(ftbzip2.c.o): In function `ft_bzip2_stream_io':
ftbzip2.c:(.text+0x358): undefined reference to `BZ2_bzDecompressEnd'
ftbzip2.c:(.text+0x3ac): undefined reference to `BZ2_bzDecompressInit'
/home/conan/w/prod-v1/BuildSingleReference/.conan/data/freetype/2.13.0/_/_/package/bf05ac88aac1986abd35f257eeb0064299824400/lib/libfreetype.a(ftbzip2.c.o): In function `FT_Stream_OpenBzip2':
ftbzip2.c:(.text+0x63a): undefined reference to `BZ2_bzDecompressInit'
collect2: error: ld returned 1 exit status
make[2]: *** [lib/libopencv_highgui.so.4.8.0] Error 1
make[1]: *** [modules/highgui/CMakeFiles/opencv_highgui.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

I guess it's an indirect issue coming from how link of ffmpeg (depends on freetype) into videoio then videoio into highgui is done in CMakeLists of OpenCV.

**Comment by SpaceIm** [comment]
> conan v1, Linux, opencv shared (and static dependencies):
> 
> ```
> /home/conan/w/prod-v1/BuildSingleReference/.conan/data/freetype/2.13.0/_/_/package/bf05ac88aac1986abd35f257eeb0064299824400/lib/libfreetype.a(sfnt.c.o): In function `sfnt_init_face':
> sfnt.c:(.text+0xc068): undefined reference to `BrotliDecoderDecompress'
> /home/conan/w/prod-v1/BuildSingleReference/.conan/data/freetype/2.13.0/_/_/package/bf05ac88aac1986abd35f257eeb0064299824400/lib/libfreetype.a(ftbzip2.c.o): In function `ft_bzip2_stream_close':
> ftbzip2.c:(.text+0x79): undefined reference to `BZ2_bzDecompressEnd'
> /home/conan/w/prod-v1/BuildSingleReference/.conan/data/freetype/2.13.0/_/_/package/bf05ac88aac1986abd35f257eeb0064299824400/lib/libfreetype.a(ftbzip2.c.o): In function `ft_bzip2_file_fill_output':
> ftbzip2.c:(.text+0x183): undefined reference to `BZ2_bzDecompress'
> /home/conan/w/prod-v1/BuildSingleReference/.conan/data/freetype/2.13.0/_/_/package/bf05ac88aac1986abd35f257eeb0064299824400/lib/libfreetype.a(ftbzip2.c.o): In function `ft_bzip2_stream_io':
> ftbzip2.c:(.text+0x358): undefined reference to `BZ2_bzDecompressEnd'
> ftbzip2.c:(.text+0x3ac): undefined reference to `BZ2_bzDecompressInit'
> /home/conan/w/prod-v1/BuildSingleReference/.conan/data/freetype/2.13.0/_/_/package/bf05ac88aac1986abd35f257eeb0064299824400/lib/libfreetype.a(ftbzip2.c.o): In function `FT_Stream_OpenBzip2':
> ftbzip2.c:(.text+0x63a): undefined reference to `BZ2_bzDecompressInit'
> collect2: error: ld returned 1 exit status
> make[2]: *** [lib/libopencv_highgui.so.4.8.0] Error 1
> make[1]: *** [modules/highgui/CMakeFiles/opencv_highgui.dir/all] Error 2
> make[1]: *** Waiting for unfinished jobs....
> ```
> 
> I guess it's an indirect issue coming from how link of ffmpeg (avfilter depends on freetype, and avfilter is a dependency of avdevice which is now really used and linked since [opencv/opencv#22706](https://github.com/opencv/opencv/pull/22706)) into videoio then videoio into highgui is done in CMakeLists of OpenCV.

I can't reproduce this issue locally in WSL.

**Comment by SpaceIm** [comment]
> > conan v1, Linux, opencv shared (and static dependencies):
> > ```
> > /home/conan/w/prod-v1/BuildSingleReference/.conan/data/freetype/2.13.0/_/_/package/bf05ac88aac1986abd35f257eeb0064299824400/lib/libfreetype.a(sfnt.c.o): In function `sfnt_init_face':
> > sfnt.c:(.text+0xc068): undefined reference to `BrotliDecoderDecompress'
> > /home/conan/w/prod-v1/BuildSingleReference/.conan/data/freetype/2.13.0/_/_/package/bf05ac88aac1986abd35f257eeb0064299824400/lib/libfreetype.a(ftbzip2.c.o): In function `ft_bzip2_stream_close':
> > ftbzip2.c:(.text+0x79): undefined reference to `BZ2_bzDecompressEnd'
> > /home/conan/w/prod-v1/BuildSingleReference/.conan/data/freetype/2.13.0/_/_/package/bf05ac88aac1986abd35f257eeb0064299824400/lib/libfreetype.a(ftbzip2.c.o): In function `ft_bzip2_file_fill_output':
> > ftbzip2.c:(.text+0x183): undefined reference to `BZ2_bzDecompress'
> > /home/conan/w/prod-v1/BuildSingleReference/.conan/data/freetype/2.13.0/_/_/package/bf05ac88aac1986abd35f257eeb0064299824400/lib/libfreetype.a(ftbzip2.c.o): In function `ft_bzip2_stream_io':
> > ftbzip2.c:(.text+0x358): undefined reference to `BZ2_bzDecompressEnd'
> > ftbzip2.c:(.text+0x3ac): undefined reference to `BZ2_bzDecompressInit'
> > /home/conan/w/prod-v1/BuildSingleReference/.conan/data/freetype/2.13.0/_/_/package/bf05ac88aac1986abd35f257eeb0064299824400/lib/libfreetype.a(ftbzip2.c.o): In function `FT_Stream_OpenBzip2':
> > ftbzip2.c:(.text+0x63a): undefined reference to `BZ2_bzDecompressInit'
> > collect2: error: ld returned 1 exit status
> > make[2]: *** [lib/libopencv_highgui.so.4.8.0] Error 1
> > make[1]: *** [modules/highgui/CMakeFiles/opencv_highgui.dir/all] Error 2
> > make[1]: *** Waiting for unfinished jobs....
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I guess it's an indirect issue coming from how link of ffmpeg (avfilter depends on freetype, and avfilter is a dependency of avdevice which is now really used and linked since [opencv/opencv#22706](https://github.com/opencv/opencv/pull/22706)) into videoio then videoio into highgui is done in CMakeLists of OpenCV.
> 
> I can't reproduce this issue locally in WSL.

@jcar87 @RubenRBS any idea?

**Comment by SpaceIm** [comment]
```
Resolved version ranges
    zlib/[>=1.2.10 <2]: zlib/1.3
Graph error
    Version conflict: Conflict between zlib/1.2.13 and zlib/1.3 in the graph.
Conflict originates from protobuf/3.21.12

Run conan graph info with your recipe and add --format=html to inspect the graph errors in an easier to visualize way.
ERROR: Version conflict: Conflict between zlib/1.2.13 and zlib/1.3 in the graph.
Conflict originates from protobuf/3.21.12

Run conan graph info with your recipe and add --format=html to inspect the graph errors in an easier to visualize way.
```

@jcar87 @RubenRBS @uilianries here is an interesting conflict with version range, quite cryptic.

**Comment by ilya-lavrenov** [comment]
> ```
> Resolved version ranges
>     zlib/[>=1.2.10 <2]: zlib/1.3
> Graph error
>     Version conflict: Conflict between zlib/1.2.13 and zlib/1.3 in the graph.
> Conflict originates from protobuf/3.21.12
> 
> Run conan graph info with your recipe and add --format=html to inspect the graph errors in an easier to visualize way.
> ERROR: Version conflict: Conflict between zlib/1.2.13 and zlib/1.3 in the graph.
> Conflict originates from protobuf/3.21.12
> 
> Run conan graph info with your recipe and add --format=html to inspect the graph errors in an easier to visualize way.
> ```
> 
> @jcar87 @RubenRBS @uilianries here is an interesting conflict with version range and conan v2 client, why fixed version doesn't have precedence to resolve conflict since it's in the range?

looks like the same issue https://github.com/conan-io/conan/issues/14007

**Comment by RubenRBS** [comment]
I think this is now ready to go, looks good to me, but could you please bump the dependencies too in this same PR? Thanks!

**Comment by SpaceIm** [comment]
Except a new version of harbuzz pushed today, there is nothing to bump. ffmpeg can't be updated, OpenCV doesn't support ffmpeg >= 5

**Comment by RubenRBS** [comment]
Added this to our next team review to help you out with that

File: recipes/opencv/4.x/conanfile.py [hunk 1]
```diff
@@ -11,7 +11,7 @@
 import re
 import textwrap
 
-required_conan_version = ">=1.54.0"
+required_conan_version = ">=1.60.0 <2.0 || >=2.0.5"
```
Comment (vobject) [inline recipes/opencv/4.x/conanfile.py:14]: Any specific reason you bumped the minimum version to 1.60?

**Comment by SpaceIm** [comment]
> > > > conan v1, Linux, opencv shared (and static dependencies):
> > > > ```
> > > > /home/conan/w/prod-v1/BuildSingleReference/.conan/data/freetype/2.13.0/_/_/package/bf05ac88aac1986abd35f257eeb0064299824400/lib/libfreetype.a(sfnt.c.o): In function `sfnt_init_face':
> > > > sfnt.c:(.text+0xc068): undefined reference to `BrotliDecoderDecompress'
> > > > /home/conan/w/prod-v1/BuildSingleReference/.conan/data/freetype/2.13.0/_/_/package/bf05ac88aac1986abd35f257eeb0064299824400/lib/libfreetype.a(ftbzip2.c.o): In function `ft_bzip2_stream_close':
> > > > ftbzip2.c:(.text+0x79): undefined reference to `BZ2_bzDecompressEnd'
> > > > /home/conan/w/prod-v1/BuildSingleReference/.conan/data/freetype/2.13.0/_/_/package/bf05ac88aac1986abd35f257eeb0064299824400/lib/libfreetype.a(ftbzip2.c.o): In function `ft_bzip2_file_fill_output':
> > > > ftbzip2.c:(.text+0x183): undefined reference to `BZ2_bzDecompress'
> > > > /home/conan/w/prod-v1/BuildSingleReference/.conan/data/freetype/2.13.0/_/_/package/bf05ac88aac1986abd35f257eeb0064299824400/lib/libfreetype.a(ftbzip2.c.o): In function `ft_bzip2_stream_io':
> > > > ftbzip2.c:(.text+0x358): undefined reference to `BZ2_bzDecompressEnd'
> > > > ftbzip2.c:(.text+0x3ac): undefined reference to `BZ2_bzDecompressInit'
> > > > /home/conan/w/prod-v1/BuildSingleReference/.conan/data/freetype/2.13.0/_/_/package/bf05ac88aac1986abd35f257eeb0064299824400/lib/libfreetype.a(ftbzip2.c.o): In function `FT_Stream_OpenBzip2':
> > > > ftbzip2.c:(.text+0x63a): undefined reference to `BZ2_bzDecompressInit'
> > > > collect2: error: ld returned 1 exit status
> > > > make[2]: *** [lib/libopencv_highgui.so.4.8.0] Error 1
> > > > make[1]: *** [modules/highgui/CMakeFiles/opencv_highgui.dir/all] Error 2
> > > > make[1]: *** Waiting for unfinished jobs....
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > I guess it's an indirect issue coming from how link of ffmpeg (avfilter depends on freetype, and avfilter is a dependency of avdevice which is now really used and linked since [opencv/opencv#22706](https://github.com/opencv/opencv/pull/22706)) into videoio then videoio into highgui is done in CMakeLists of OpenCV.
> > > 
> > > 
> > > I can't reproduce this issue locally in WSL.
> > 
> > 
> > @jcar87 @RubenRBS any idea?
> 
> Still no idea what's wrong here, any help would be appreciated. Error only in 4.8.0, and I don't know why there are undefined references. I've checked source code and it doesn't make sense to have this error, CMake target of freetype is linked so transitive libs should be linked.

I've tested locally on Ubuntu 22.04 and GCC 13, OpenCV shared, and dependencies static, so identical to build in failure in c3i, and I can't reproduce this issue.

freetype comes from ffmpeg, optional dependency of opencv_videoio. Actually freetype is not even propagated by ffmpeg components opencv depends on, so I don't understand why there is even a reference to freetype in link commands during this build. Since log of v1 pipeline is messed up (errors are concatenated at the end, we can't understand the context), I can't understand what's wrong here.

**Comment by SpaceIm** [comment]
So no link error for other versions, it's specific to 4.8.x

**Comment by SpaceIm** [comment]
```
/usr/bin/g++ -fPIC -m64   -fsigned-char -W -Wall -Wreturn-type -Wnon-virtual-dtor -Waddress -Wsequence-point -Wformat -Wformat-security -Wmissing-declarations -Wundef -Winit-self -Wpointer-arith -Wshadow -Wsign-promo -Wuninitialized -Wsuggest-override -Wno-delete-non-virtual-dtor -Wno-comment -fdiagnostics-show-option -Wno-long-long -pthread -fomit-frame-pointer -ffunction-sections -fdata-sections  -msse -msse2 -msse3 -fvisibility=hidden -fvisibility-inlines-hidden -Wno-deprecated-declarations -O3 -DNDEBUG  -DNDEBUG -m64  -Wl,--gc-sections -Wl,--as-needed -Wl,--no-undefined -shared -Wl,-soname,libopencv_highgui.so.408 -o ../../lib/libopencv_highgui.so.4.8.1 CMakeFiles/opencv_highgui.dir/src/backend.cpp.o CMakeFiles/opencv_highgui.dir/src/window.cpp.o CMakeFiles/opencv_highgui.dir/src/roiSelector.cpp.o CMakeFiles/opencv_highgui.dir/src/window_gtk.cpp.o  -Wl,-rpath,/home/conan/w/prod-v1/bsr/44621/bddac/.conan/data/opencv/4.8.1/_/_/build/81964eb8d1736c4bf89840fae039bc86f45d1113/build/Release/lib: ../../lib/libopencv_videoio.so.4.8.1 -ldl -lm -lpthread -lrt ../../lib/libopencv_imgcodecs.so.4.8.1 ../../lib/libopencv_imgproc.so.4.8.1 ../../lib/libopencv_core.so.4.8.1 -lm /usr/lib/x86_64-linux-gnu/libgtk-x11-2.0.so /usr/lib/x86_64-linux-gnu/libgdk-x11-2.0.so /usr/lib/x86_64-linux-gnu/libpangocairo-1.0.so /usr/lib/x86_64-linux-gnu/libatk-1.0.so /usr/lib/x86_64-linux-gnu/libcairo.so /usr/lib/x86_64-linux-gnu/libgdk_pixbuf-2.0.so /usr/lib/x86_64-linux-gnu/libgio-2.0.so /usr/lib/x86_64-linux-gnu/libpangoft2-1.0.so /usr/lib/x86_64-linux-gnu/libpango-1.0.so /usr/lib/x86_64-linux-gnu/libgobject-2.0.so /usr/lib/x86_64-linux-gnu/libfontconfig.so /home/conan/w/prod-v1/bsr/44621/bddac/.conan/data/freetype/2.13.0/_/_/package/bf05ac88aac1986abd35f257eeb0064299824400/lib/libfreetype.a /usr/lib/x86_64-linux-gnu/libgthread-2.0.so /usr/lib/x86_64-linux-gnu/libglib-2.0.so 
modules/highgui/CMakeFiles/opencv_highgui.dir/build.make:168: recipe for target 'lib/libopencv_highgui.so.4.8.1' failed
make[2]: Leaving directory '/home/conan/w/prod-v1/bsr/44621/bddac/.conan/data/opencv/4.8.1/_/_/build/81964eb8d1736c4bf89840fae039bc86f45d1113/build/Release'
CMakeFiles/Makefile2:1572: recipe for target 'modules/highgui/CMakeFiles/opencv_highgui.dir/all' failed
```

Ooooh, a beautiful side effect of gtk as a system lib. Indeed gtk depends on freetype, but it depends on system freetype. For some reason, package_info() of gtk finds conan flavor of freetype (dragged by ffmpeg): https://github.com/conan-io/conan-center-index/blob/6a0354502df1c010acc31d346eb4c22188b37d08/recipes/gtk/system/conanfile.py#L50-L53

And since freetype is static in conan dependency graph, and this package_info() of gtk just relies on system pc file of gtk, it cannot model transitive dependencies (because it shouldn't be there from system lib point of view).

**Comment by SpaceIm** [comment]
```
ERROR: Conflict in opencv/4.8.1:
    'opencv/4.8.1' requires 'wayland-protocols/1.32' while 'xkbcommon/1.6.0' requires 'wayland-protocols/1.31'.
    To fix this conflict you need to override the package 'wayland-protocols' in your root package.
```

This `wayland-protocols` in requirements depending on whether consumer use 1 or 2 profiles should be removed in `xkbcommon`

**Comment by SpaceIm** [comment]
Ok now it seems to pass v1 pipeline. There is a missing package of vulkan-loader in v2 pipeline after a small change in vulkan-headers recipe. https://github.com/conan-io/conan-center-index/pull/20899 should fix this.

**Comment by SpaceIm** [comment]
@uilianries @RubenRBS could you review this one please? It has been a long time and it finally builds.

I had to switch to wayland (instead of gtk) for the backend of highgui component of OpenCV >= 4.8.0 because this recipe requires `gtk/system` which leads to incoherency/collision between system dependencies of gtk (freetype etc) and dependency graph of opencv recipe.

**Review by uilianries — APPROVED** [review]
- Flatbuffers: https://github.com/opencv/opencv/blob/4.8.1/CMakeLists.txt#L476
- Protobuf: https://github.com/opencv/opencv/blob/4.8.1/CMakeLists.txt#L437
- Wayland: https://github.com/opencv/opencv/blob/4.8.1/CMakeLists.txt#L289
- Avif: https://github.com/opencv/opencv/blob/4.8.1/CMakeLists.txt#L243

**Review by ilya-lavrenov — APPROVED** [review]

**Review by jwillikers — APPROVED** [review]

**Pull request merged**